src/data/java_profile.rs

```diff
@@ -126,15 +74,112 @@ impl CollectData for JavaProfileRaw {
+            return Some(key.to_string());
+        }
+
+        self.process_map.iter().find_map(|(jid, name)| {
+            if key == name[0] {
+                Some(jid.to_string())
+            } else {
+                None
+            }
+        })
     }
 
-    fn finish_data_collection(&mut self, _params: CollectorParams) -> Result<()> {
+    fn update_process_map(&mut self) -> String {
+        info!("Running jps (may incur utilization spike)...");
+        let jps_cmd = Command::new("jps").output();
+        let mut jps_str = String::new();
+        match jps_cmd {
+            Ok(jps_out) => {
+                jps_str = String::from_utf8(jps_out.stdout).unwrap_or_default();
+                let jps: Vec<&str> = jps_str.split_whitespace().collect();
+                for i in (0..jps.len()).step_by(2) {
+                    if jps[i + 1] != "Jps" {
+                        self.process_map
+                            .insert(String::from(jps[i]), vec![String::from(jps[i + 1])]);
+                    }
+                }
+            }
+            Err(e) => {
+                error!("Jps command failed: {}", e);
+            }
+        }
+        jps_str
+    }
+}
+
+impl CollectData for JavaProfileRaw {
+    fn prepare_data_collector(&mut self, params: CollectorParams) -> Result<()> {
+        let jps_str = self.update_process_map();
+        let jps: Vec<&str> = jps_str.split_whitespace().collect();
+
+        let mut jids: Vec<String> = Vec::new();
+        let default = String::from("jps");
+        let jprofile = params
+            .profile
+            .get(JAVA_PROFILE_FILE_NAME)
+            .unwrap_or(&default)
+            .as_str();
+        match jprofile {
+            "jps" => jids = self.process_map.clone().into_keys().collect(),
+            _ => {
+                let args: Vec<&str> = jprofile.split(',').collect();
+                for arg in args {
+                    if !jps.contains(&arg) {
+                        error!("No JVM with name/PID '{}'.", arg);
+                        continue;
+                    } else if jps.iter().position(|&r| r == arg).unwrap()
+                        != jps.iter().rposition(|&r| r == arg).unwrap()
+                    {
+                        error!("Multiple JVMs with the name '{}', please provide PID.", arg);
+                        continue;
+                    }
+                    if let Some(jid) = self.get_jid(arg) {
+                        jids.push(jid);
+                    }
+                }
+            }
+        }
+
+        self.launch_asprof(jids, 0, params)
+    }
+
+    fn collect_data(&mut self, params: CollectorParams) -> Result<()> {
+        let jprofile = params.profile.get(JAVA_PROFILE_FILE_NAME).unwrap().as_str();
+        if jprofile != "jps" {
+            return Ok(());
+        }
+
+        static COUNTER: AtomicU64 = AtomicU64::new(0);
+        COUNTER.fetch_add(params.collection_interval, Ordering::Relaxed);
```

Comment:
The collection_interval is unlikely to change. Why do we need Atomic operations for it? Is the collection_interval the time elapsed till now?

Reply:
Yes, the counter is the current time elapsed so when we launch a new async-profiler in the middle of the record we know what time to set.